Split module into multiple files

Lucki opened the pull request:
The main script is really long with all the functionality we've added.
I've tried to split it into multiple files related what the function are doing:
* init is handling starting the mod and creating the arg_parser
* downloads contains everything related fetching from the net
* utils contains helper functions like waiting for steam or retrieving local information
* variables has all definitions combined

I haven't tested this yet, I want to hear your opinion first. Is that something useful? Is that something which is common in python?

kakurasan commented:
> The main script is really long with all the functionality we've added.

> I haven't tested this yet, I want to hear your opinion first. Is that something useful? Is that something which is common in python?

I think it's better to split `__init__.py` (43.8KB) and the work will *enhance readability*.

Many Python packages provide multiple *.py files.

Go ahead and do it :)

kakurasan commented on .github/workflows/check_main_script.yaml:
```diff
@@ -31,5 +37,16 @@ jobs:
         run: pip install flake8 flake8-docstrings flake8-import-order
       - name: Check main script (bin) by flake8
         run: flake8 --max-line-length=90 --show-source --statistics truckersmp-cli
+        if: always()
       - name: Check main script (module) by flake8
         run: flake8 --max-line-length=90 --show-source --statistics truckersmp_cli/__init__.py
+        if: always()
+      - name: Check downloads script (module) by flake8
+        run: flake8 --max-line-length=90 --show-source --statistics truckersmp_cli/downloads.py
+        if: always()
+      - name: Check utils script (module) by flake8
+        run: flake8 --max-line-length=90 --show-source --statistics truckersmp_cli/utils.py
+        if: always()
+      - name: Check variables script (module) by flake8
+        run: flake8 --max-line-length=90 --show-source --statistics truckersmp_cli/variables.py
+        if: always()
```
`flake8` can specify multiple files.

Currently (in master branch) we have 2 steps for running `flake8` (1 step for `truckersmp-cli`, 1 step for `__init__.py`) but now I think we can check all the scripts in 1 step:

```suggestion
      - name: Check Python scripts by flake8
        run: flake8 --max-line-length=90 --show-source --statistics truckersmp-cli truckersmp_cli/*.py
```

When `flake8` fails, we can see the error details (errored file, location, and error code/error description).

kakurasan commented on .github/workflows/check_main_script.yaml:
```diff
@@ -17,10 +20,13 @@ on:
       - .github/workflows/check_main_script.yaml
       - truckersmp-cli
       - truckersmp_cli/__init__.py
+      - truckersmp_cli/downloads.py
+      - truckersmp_cli/utils.py
+      - truckersmp_cli/variables.py
```
```suggestion
      - truckersmp_cli/*.py
```

kakurasan commented:
Actually, I'd not like to use `filename.functionname` (e.g. `utils.check_steam_process()`) or `filename.ClassName` (e.g. `variables.URL.release`) for smaller code size.
`from truckersmp_cli.variables import Dir, URL` (or `from .variables import Dir, URL`) allows us to write `Dir.default_moddir`, `URL.dlurl`, etc.

However, the current code won't work because of *circular imports* even if specifying `from truckersmp_cli.[filename] import [name...]` (or `from .[filename] import [name...]`).

I think we need to *change how to split* to fix the circular dependencies.

I guess that the functions in `downloads.py` should be part of `utils.py` and we should add a file (`htmlparsers.py` or something else) for `DowngradeHTMLParser`.

Lucki commented:
I've moved the functions a bit around and split them further apart.
* args.py: Related to program arguments handling
* __init__.py: Main functions
* steamcmd.py: Related to steamcmd
* truckersmp.py: Things related to interacting with TruckersMP
* utils.py: Helper functions
* variables.py

Only thing I can imagine being a circular dependency later on is `from .truckersmp import download_files` in `utils.py`. Everything other should be resolved.
I gave it some quick tests with updating and starting through proton and it seems to work as expected.

Lucki commented:
> I think `download_files()` should be part of `utils.py`

Right, I moved it there.

kakurasan commented:
1. Don't change the arguments and calls for `activate_native_d3dcompiler_47()` *in this PR*.
2. I think the current code in `__init__.py` can be moved to a new file (e.g. `main.py`):

* Move the current code in `__init__.py` to `main.py`
* Leave `__init__.py` empty (Only docstring is needed for `flake8`. Don't remove the file, our package won't work without this file)
* Replace `from truckersmp_cli import main` with `from truckersmp_cli.main import main` in the main script `truckersmp-cli`

Lucki commented:
> 1. Don't change the arguments and calls for `activate_native_d3dcompiler_47()` _in this PR_.

But without that it can't access `args` anymore. My intention was to keep it working.

kakurasan commented on truckersmp_cli/args.py:
```diff
@@ -0,0 +1,237 @@
+from .variables import AppId, Dir
+
+
+def check_args_errors(args):
+    """Check command-line arguments."""
```
```suggestion
def check_args_errors():
    """Check command-line arguments."""
    global args

```

kakurasan commented on truckersmp_cli/steamcmd.py:
```diff
@@ -0,0 +1,181 @@
+from .variables import Dir, URL
+
+
+def update_game(args):
+    """
+    Update game and Proton via SteamCMD.
+
+    We make sure Steam is closed before updating.
+    On Linux, we make sure both Windows and Linux version of Steam are closed.
+    It's possible to update with the Steam client open but the client looses
+    all connectivity and asks for password and Steam Guard code after restart.
+
+    When "--wine" is specified, this function retrieves/uses Windows version of
+    SteamCMD. When "--proton" is specified, this retrieves/uses
+    Linux version of SteamCMD.
+    """
```
```suggestion
def update_game():
    """
    Update game and Proton via SteamCMD.

    We make sure Steam is closed before updating.
    On Linux, we make sure both Windows and Linux version of Steam are closed.
    It's possible to update with the Steam client open but the client looses
    all connectivity and asks for password and Steam Guard code after restart.

    When "--wine" is specified, this function retrieves/uses Windows version of
    SteamCMD. When "--proton" is specified, this retrieves/uses
    Linux version of SteamCMD.
    """
    global args

```

kakurasan commented on truckersmp_cli/utils.py:
```diff
@@ -0,0 +1,405 @@
+
+
+def activate_native_d3dcompiler_47(prefix, wine, ets2):
+    """Download/activate native 64-bit version of d3dcompiler_47."""
+    # check whether DLL is already downloaded
+    md5hash = hashlib.md5()
+    need_download = True
+    try:
+        with open(File.d3dcompiler_47, "rb") as f:
+            while True:
+                buf = f.read(md5hash.block_size * 4096)
+                if not buf:
+                    break
+                md5hash.update(buf)
+        if md5hash.hexdigest() == File.d3dcompiler_47_md5:
+            logging.debug("d3dcompiler_47.dll is present, MD5 is OK.")
+            need_download = False
+    except Exception:
+        pass
+
+    # download 64-bit d3dcompiler_47.dll from ImagingSIMS' repo
+    # https://github.com/ImagingSIMS/ImagingSIMS
+    if need_download:
+        logging.debug("Downloading d3dcompiler_47.dll")
+        os.makedirs(Dir.dllsdir, exist_ok=True)
+        if not download_files(
+          URL.raw_github,
+          [(URL.d3dcompilerpath, File.d3dcompiler_47, File.d3dcompiler_47_md5), ]):
+            sys.exit("Failed to download d3dcompiler_47.dll")
+
+    # copy into system32
+    destdir = os.path.join(prefix, "drive_c/windows/system32")
+    logging.debug("Copying d3dcompiler_47.dll into {}".format(destdir))
+    shutil.copy(File.d3dcompiler_47, destdir)
+
+    # add DLL override setting
+    env = os.environ.copy()
+    env["WINEDEBUG"] = "-all"
+    env["WINEPREFIX"] = prefix
+    exename = "eurotrucks2.exe" if ets2 else "amtrucks.exe"
```
```suggestion
def activate_native_d3dcompiler_47(prefix, wine):
    """Download/activate native 64-bit version of d3dcompiler_47."""
    global args

    # check whether DLL is already downloaded
    md5hash = hashlib.md5()
    need_download = True
    try:
        with open(File.d3dcompiler_47, "rb") as f:
            while True:
                buf = f.read(md5hash.block_size * 4096)
                if not buf:
                    break
                md5hash.update(buf)
        if md5hash.hexdigest() == File.d3dcompiler_47_md5:
            logging.debug("d3dcompiler_47.dll is present, MD5 is OK.")
            need_download = False
    except Exception:
        pass

    # download 64-bit d3dcompiler_47.dll from ImagingSIMS' repo
    # https://github.com/ImagingSIMS/ImagingSIMS
    if need_download:
        logging.debug("Downloading d3dcompiler_47.dll")
        os.makedirs(Dir.dllsdir, exist_ok=True)
        if not download_files(
          URL.raw_github,
          [(URL.d3dcompilerpath, File.d3dcompiler_47, File.d3dcompiler_47_md5), ]):
            sys.exit("Failed to download d3dcompiler_47.dll")

    # copy into system32
    destdir = os.path.join(prefix, "drive_c/windows/system32")
    logging.debug("Copying d3dcompiler_47.dll into {}".format(destdir))
    shutil.copy(File.d3dcompiler_47, destdir)

    # add DLL override setting
    env = os.environ.copy()
    env["WINEDEBUG"] = "-all"
    env["WINEPREFIX"] = prefix
    exename = "eurotrucks2.exe" if args.ets2 else "amtrucks.exe"
```

kakurasan commented on truckersmp_cli/main.py:
```diff
@@ -0,0 +1,241 @@
+    # activate native d3dcompiler_47
+    wine = os.path.join(args.protondir, "dist/bin/wine")
+    if args.activate_native_d3dcompiler_47:
+        activate_native_d3dcompiler_47(
+            os.path.join(args.prefixdir, "pfx"),
+            wine,
+            args.ets2)
```
```suggestion
        activate_native_d3dcompiler_47(os.path.join(args.prefixdir, "pfx"), wine)
```

kakurasan commented on truckersmp_cli/main.py:
```diff
@@ -0,0 +1,241 @@
+    logging.info("Mod directory: " + args.moddir)
+
+    # check for errors
+    check_args_errors(args)
```
```suggestion
    check_args_errors()
```

kakurasan commented on truckersmp_cli/main.py:
```diff
@@ -0,0 +1,241 @@
+    # download/update ATS/ETS2 and Proton
+    if args.update:
+        logging.debug("Updating game files")
+        update_game(args)
```
```suggestion
        update_game()
```

kakurasan commented on truckersmp_cli/main.py:
```diff
@@ -0,0 +1,241 @@
+    """Start game with Wine."""
+    wine = os.environ["WINE"] if "WINE" in os.environ else "wine"
+    if args.activate_native_d3dcompiler_47:
+        activate_native_d3dcompiler_47(args.prefixdir, wine, args.ets2)
```
```suggestion
        activate_native_d3dcompiler_47(args.prefixdir, wine)
```

kakurasan commented on truckersmp_cli/steamcmd.py:
```diff
@@ -0,0 +1,181 @@
+"""
+steamcmd handler for truckersmp-cli main script.
```
```suggestion
SteamCMD handler for truckersmp-cli main script.
```

kakurasan commented:
1. Create a new class named `Args` in `variables.py`
```python
class Args:
    """Arguments from command line."""

    pass
```
2. Replace existing `args.` with `Args.`
3. Import `Args` in the files that use the values in `args` (e.g. `from .variables import AppId, Args, Dir, File`)
4. Remove `global args` from `main()`
5. Replace `args = arg_parser.parse_args()` with `arg_parser.parse_args(namespace=Args)` in `main()`
6. Undo editing `check_args_errors()`, `update_game()`, and `activate_native_d3dcompiler_47()`

Lucki commented:
My pylint is now going bonkers… but I'm no python specialist as you may already know.

Lucki commented:
> I think we can ignore/suppress them.

ok

kakurasan commented on truckersmp_cli/utils.py:
```diff
@@ -0,0 +1,405 @@
+
+    This function depends on the package "vdf".
+    """
+    if not vdf_is_available:
+        return None
+
```
Sorry, I missed this:
```suggestion
```
This check was added in this PR. Do you have any reason to do the double check for the splitting?

See also https://github.com/lhark/truckersmp-cli/pull/67#discussion_r429602008

Lucki replied:
No, also forgot about this one. But in retrospect it couldn't hurt to include a check where it actually could fail.

kakurasan commented on truckersmp_cli/args.py:
```diff
@@ -0,0 +1,237 @@
+      "-i", "--proton-appid", metavar="APPID", type=int,
+      default=AppId.proton[AppId.proton["default"]],
+      help="""choose a different AppID for Proton (Needs an update for changes)
+              [Default:{}] """.format(AppId.proton[AppId.proton["default"]]))
```
I missed this too.
```suggestion
              [Default: {}]""".format(AppId.proton[AppId.proton["default"]]))
```
Was the space moved accidentally (or intentionally)?

Lucki replied:
That was my IDE thinking better. I've turned of "formatting on save" before because of this but "formatting on paste" was still on… besides that I'm unsure why it's formatting in the first place because there isn't a formatting prover chosen for python. Anyway, I turned it off, thanks for spotting.
